src/Compilers/CSharp/Portable/Symbols/Source/ModifierUtils.cs

```diff
@@ -417,6 +422,21 @@ internal static CSDiagnosticInfo CheckAccessibility(DeclarationModifiers modifie
                 }
             }
 
+            if ((modifiers & DeclarationModifiers.Required) != 0)
+            {
+                switch (symbol)
+                {
+                    case FieldSymbol or PropertySymbol when symbol.DeclaredAccessibility < symbol.ContainingType.DeclaredAccessibility:
+                    case PropertySymbol { SetMethod.DeclaredAccessibility: var accessibility } when accessibility < symbol.ContainingType.DeclaredAccessibility:
+                        // Required member '{0}' cannot be less visible or have a setter less visible than the containing type '{1}'.
+                        return new CSDiagnosticInfo(ErrorCode.ERR_RequiredMemberCannotBeLessVisibleThanContainingType, symbol, symbol.ContainingType);
+                    case PropertySymbol { SetMethod: null }:
+                    case FieldSymbol when (modifiers & DeclarationModifiers.ReadOnly) != 0:
```

Comment:
There's also a `readonly` scenario with properties, although it has a slightly different meaning.

Reply:
> There's also a readonly scenario with properties, although it has a slightly different meaning.

That change in meaning is enough that it shouldn't affect required. You can have a `readonly` setter.